src/Microsoft.AspNetCore.Razor.Evolution/DirectiveDescriptorBuilder.cs

```diff
@@ -35,39 +37,58 @@ public DefaultDirectiveDescriptorBuilder(string name, DirectiveDescriptorKind ty
                 _tokenDescriptors = new List<DirectiveTokenDescriptor>();
             }
 
-            public IDirectiveDescriptorBuilder AddType()
+            public IDirectiveDescriptorBuilder AddType(bool optional)
```

Comment:
Instead of having each API specify whether its optional or not we could make an API that direct the user into a state where they can't fail. Aka:

```
builder
    .AddType()
    .RemainingTokensOptional()
    .AddString()
    .AddMember();
```

And

```
builder
    .AddType()
    .RemainingTokensOptional()
    .AddString()
    .RemainingTokensOptional() // Explode
    .AddMember();
```

We all know I'm horrible at naming choices but you get the gist 😄.